### PR TITLE
Separate querybuilder for navigation and search

### PR DESCRIPTION
**Pull request by zep-com**

### CODE REVIEW

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    private $installerService;
+
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $this->config = Shopware()->Container()->get('config');
+        $this->installerService = Shopware()->Container()->get('shopware_plugininstaller.plugin_manager');
+        Shopware()->Session()->offsetSet('isSearchPage', false);
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: Implement a tearDown method (should be placed right after setUp) which unsets `isSearchPage`.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        $shopUrl = rtrim(Shopware()->Shop()->getHost(), '/');
+        $parameters = [
+            'outputAdapter' => 'XML_2.0',
+            'userip' => 'UNKNOWN',
+            'revision' => $plugin->getVersion(),
+            'shopkey' => $this->config->offsetGet('ShopKey')
+        ];
+        $aliveUrl = sprintf(
+            '%s/%s/%s?shopkey=%s',
+            QueryBuilder::BASE_URL,
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: I would prefer that you use the actual value instead of the constant. Which also means that you don't have to provide it as parameter but can directly put it into the string.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            $shopUrl,
+            QueryBuilder::ALIVE_ENDPOINT,
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: I would prefer that you use the actual value instead of the constant. Which also means that you don't have to provide it as parameter but can directly put it into the string.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            $parameters['shopkey']
+        );
+        $executeUrl = sprintf(
+            '%s/%s/%s?%s',
+            QueryBuilder::BASE_URL,
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: I would prefer that you use the actual value instead of the constant. Which also means that you don't have to provide it as parameter but can directly put it into the string.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            $shopUrl,
+            NavigationQueryBuilder::ENDPOINT,
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: I would prefer that you use the actual value instead of the constant. Which also means that you don't have to provide it as parameter but can directly put it into the string.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            http_build_query($parameters)
+        );
+        $httpResponse = new Response(200, [], 'alive');
+        $httpClientMock = $this->createMock(GuzzleHttpClient::class);
+        $aliveCallback = $this->returnCallback(function ($url) use ($aliveUrl, $httpResponse) {
+            Assert::assertSame($aliveUrl, $url);
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: Since we are checking the provided parameter, this assertion should be made in the `with` part. So check if something like this is possible:
```
$httpClientMock->expects($this->exactly(2))
    ->method('get')
    ->with($this->onConsecutive([THE CALLBACKS GOE HERE]))
    ->will($this->onConsecutiveCalls($httpResponse, $httpResponse));
```
See  `\FinSearchUnified\Tests\Subscriber\FrontendTest::testBeforeSearchIndexAction` for reference.

If it turns out that this approach won't work, we can stick with the current implementation.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php [hunk 7]
```diff
@@ -278,6 +290,22 @@ public function testCreateProductQueryWithOffsetAndLimit($offset, $expectedOffse
...
+     * @param string $expectedInstance
+     *
+     * @throws Exception
+     */
+    public function testTypeOfQueryBuilder($isSearchPage, $expectedInstance)
+    {
+        Shopware()->Session()->offsetSet('isSearchPage', $isSearchPage);
+
+        $builder = $this->factory->createQueryBuilder();
+        $this->assertInstanceOf($expectedInstance, $builder, sprintf('Expected %s to be returned', $expectedInstance));
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php]: The third parameter won't be necessary since `assertInstanceOf` will already give you a similar message.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php [hunk 8]
```diff
@@ -302,4 +330,12 @@ public function offsetLimitProvider()
             'Offset is 5 and Limit = 2' => [5, 5, 2, 2],
         ];
     }
+
+    public function isSearchPageDataProvider()
+    {
+        return [
+            'isSearchPage is true' => [true, SearchQueryBuilder::class],
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php]: Rename the data sets to `Search request` and `Navigation request`.

File: FinSearchUnified/Tests/Subscriber/WidgetsTest.php [hunk 9]
```diff
@@ -11,12 +11,12 @@
 
 class WidgetsTest extends TestCase
 {
-    protected static $ensureLoadedPlugins = [
-        'FinSearchUnified' => [
-            'ShopKey' => '0000000000000000ZZZZZZZZZZZZZZZZ',
-            'ActivateFindologic' => true
-        ],
-    ];
+    public static function setUpBeforeClass()
```
Comment (ihanli) [inline FinSearchUnified/Tests/Subscriber/WidgetsTest.php]: Why was this change necessary? `$ensureLoadedPlugins` should already have worked (at least theoretically).

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php [hunk 10]
```diff
@@ -66,23 +65,15 @@ public function __construct(
...
-        } else {
-            $endpoint = self::NAVIGATION_ENDPOINT;
-        }
-
         $url = sprintf(
             '%s/%s/%s?%s',
             self::BASE_URL,
             $this->shopUrl,
-            $endpoint,
+            static::ENDPOINT,
```
Comment (ihanli) [inline FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: Are we sure that `self::ENDPOINT` won't work here?

File: FinSearchUnified/Tests/Subscriber/WidgetsTest.php (hunk 9 above)
Comment (zep-com) [reply FinSearchUnified/Tests/Subscriber/WidgetsTest.php]: Since we have disabled Findologic by default in the last version, it did not get activated by using `ensureLoadedPlugins`'s value after a fresh setup of the environment. Not sure why though.

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php (hunk 10 above)
Comment (zep-com) [reply FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: Yes it was getting the Querybuilder's `index` endpoint, not the child class's overridden endpoint.

File: FinSearchUnified/Tests/Subscriber/WidgetsTest.php (hunk 9 above)
Comment (ihanli) [reply FinSearchUnified/Tests/Subscriber/WidgetsTest.php]: Well, that's unfortunate.

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php (hunk 10 above)
Comment (ihanli) [reply FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: I checked the PHP manual (just because I was curious). It seems that constants are definitely not statics but you have to access it this way since the constant is shared between multiple classes (QueryBuilder and Navigation-/SearchQueryBuilder).

The more you know.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php (hunk 6 above)
Comment (ihanli) [reply FinSearchUnified/Tests/Bundle/SearchBundleFindologic/NavigationQueryBuilderTest.php]: Well, that's unfortunate but okay.
